### ISSUE TITLE
Hackathon Registration Process

### ISSUE THREAD

**Issue by rickkdev** [OpenQDev/OpenQ-Frontend#938]
### 📃 Description

We want to enable a registration process for each hackathon. 

A hackathon is a repository, challenges are issues.

I.e. if I want to create a Hackathon, I have to create my first Issue within a repository. We then give the option to add a registration process if desired, which only maintainers of this repository can oversee (stats on how many have registered can then be viewed).

Users register by filling out a form on our website which is auto-filled when the user logs in or is already logged in or have to be filled out manually as a guest.

### 👨‍⚖️ Acceptance Criteria

- [ ] If a user creates a new contest for a repo that has no registration process yet we show in the model additional info (enable the hackathon registration process
  - [x] this will add new input fields where you can set the time when the hackathon will start and at what time the deadline is for the registration
  - [ ] if a contest is created with the registration process the contest/ org is especially highlighted in our contest explorer, more specific mocks for that coming tomorrow
  - [ ] user can then go to the website and explore contests/ hackathons that are coming up soon and register with a simple button click
  - [ ] if they are logged in we show them an auto-filled form with all information we got from their profile and they can add more stuff to it (which we also store for the profile page permanently 
  - [ ] if they are not logged in we ask them to either login or register as guests which would mean we ask them for specific information and do a soft registration (more in the mock)
  - [ ] after registration the process is completed and they get via email more updates about the hackathon
  - [ ] the maintainer can go on the hackathon and see how many people signed up check out their profile and export their emails to import them in their favorite email mass messenger (Mailchimp or whatever)

**Comment by FlacoJones** [comment]
We don't have any Issue<>Repository linked relationship at the contract or subgraph level, so this would have to be added to the API and related there.

**Comment by FlacoJones** [comment]
```
model Contest {
  id                String         @id @map("_id") // this would be the repository node ID in the Github API
  bountyIds       String[]
  bounties        Bounty[]       @relation(fields: [bountyIds], references: [address])
  participantIds       String[]
  participants        User[]       @relation(fields: [participantIds], references: [address])
  organization Organization @relation(fields: [organizationIds], references: [id])
  organizationId String @db.ObjectId
}
```

**Comment by FlacoJones** [comment]
A "bounty" in this case is just the challenge basically

**Comment by FlacoJones** [comment]
Upon minting the first issue in the repo, we could use `OpenQ-Bounty-Actions` to create the new Contest model in the API

**Comment by FlacoJones** [comment]
The above model would be the bare minimum, and then what the admin (inherit the auth from Github? or have it be the wallet of the admin has edit access?) could add is like start time, deadline etc. as optional fields

**Comment by Christopher-Stevers** [comment]
How are participants going to be directed to register with the hackathon. I understand they can view them without registering, but we should probably have a link that can be sent to participants that brings them directly to the registration page.

**Comment by Christopher-Stevers** [comment]
Ideally this page would funnel them through signing for the first time with github and registering for the hackathon, as well as exposing them to any upcoming hackathons.

**Comment by FlacoJones** [comment]
Right, ya we should have a shareable link like `/registration/repositoryId` that let's them do this.

By "signing up" we are essentially adding them to the Contests [participants] array

**Comment by FlacoJones** [comment]
I suggest `repositoryId` since each repo is a hackathon
